### PR TITLE
Player aggro division 

### PR DESCRIPTION
**Pull request by RobbeBryssinck**
This should make it so that enemies targets the entire party more evenly. An enemy will target whichever player has hit them last.

### CODE REVIEW

File: Code/client/Services/CombatService.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+struct ProjectileLaunchedEvent;
+struct NotifyProjectileLaunch;
+struct HitEvent;
+
+/**
+* @brief Responsible for projectiles, combat agro, etc.
+*/
+struct CombatService
+{
+public:
```
Comment (yamashi) [inline Code/client/Services/CombatService.h:15]: not needed

File: Code/client/Services/Generic/CombatService.cpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    const uint32_t cSpellId = modSystem.GetGameId(acMessage.SpellID);
+    launchData.pSpell = Cast<MagicItem>(TESForm::GetById(cSpellId));
+
+    launchData.eCastingSource = (MagicSystem::CastingSource)acMessage.CastingSource;
+
+    launchData.iArea = acMessage.Area;
+    launchData.fPower = acMessage.Power;
+    launchData.fScale = acMessage.Scale;
+
+    launchData.bAlwaysHit = acMessage.AlwaysHit;
```
Comment (yamashi) [inline Code/client/Services/Generic/CombatService.cpp:164]: Could be better to name them the same?

File: Code/client/Games/Skyrim/Actor.h [hunk 3]
```diff
@@ -234,11 +234,18 @@ struct Actor : TESObjectREFR
     void Kill() noexcept;
     void Reset() noexcept;
     void Respawn() noexcept;
+
     void PickUpObject(TESObjectREFR* apObject, int32_t aCount, bool aUnk1, float aUnk2) noexcept;
     void DropObject(TESBoundObject* apObject, ExtraDataList* apExtraData, int32_t aCount, NiPoint3* apLocation, NiPoint3* apRotation) noexcept;
     void DropOrPickUpObject(const Inventory::Entry& arEntry, NiPoint3* apPoint, NiPoint3* apRotate) noexcept;
     void SpeakSound(const char* pFile);
 
+    bool IsInCombat() noexcept;
```
Comment (yamashi) [inline Code/client/Games/Skyrim/Actor.h:243]: const [[nodiscard]]

File: Code/client/Games/Skyrim/Actor.h [hunk 4]
```diff
@@ -234,11 +234,18 @@ struct Actor : TESObjectREFR
...
+    Actor* GetCombatTarget() noexcept;
```
Comment (yamashi) [inline Code/client/Games/Skyrim/Actor.h:244]: const [[nodiscard]]

**Review by yamashi — CHANGES_REQUESTED** [review]
Nice job, very minor changes

File: Code/client/Services/Generic/CombatService.cpp (hunk 2 above)
Comment (RobbeBryssinck) [reply Code/client/Services/Generic/CombatService.cpp:164]: I considered that, but the member names in LaunchData are closer to the actual names used in the in-game struct. Could do either way.

**Comment by kulcris** [comment]
I am curious if it will cause target swapping a bit too much though. resulting in playing a game of keep away with a single enemy. 
Could this be adjusted to be the highest damaging player instead?

**Comment by RobbeBryssinck** [comment]
> I am curious if it will cause target swapping a bit too much though. resulting in playing a game of keep away with a single enemy. Could this be adjusted to be the highest damaging player instead?

Yes, and I did design it with that in mind, so that the enemy is locked onto one player for 5 seconds, but I see now that there's a bug and that that timer is ignored. I will fix this tomorrow.

**Pull request merged**